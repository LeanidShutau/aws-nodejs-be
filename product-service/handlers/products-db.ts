import { Client } from 'pg';
const { PG_HOST, PG_PORT, PG_DATABASE, PG_USERNAME, PG_PASSWORD } = process.env;
const dbOptions = {
  host: PG_HOST,
  port: Number(PG_PORT),
  database: PG_DATABASE,
  user: PG_USERNAME,
  password: PG_PASSWORD,
  ssl: {
    rejectUnauthorized: false,
  },
  connectionTimeoutMillis: 5000,
};

export interface ProductEntity {
  id: string,
  title: string,
  description: string,
  price: number,
  image: string,
  count: number,
}

export interface StockEntity {
  product_id: string,
  count: number,
}

export async function findAll(): Promise<ProductEntity[]> {
  const client = new Client(dbOptions);
  await client.connect();
  try {
    const productsResponse = await client.query(`
      select
        PR.*,
        coalesce(ST.count, 0) as count
      from
        public.products PR
      left join public.stocks ST on
        PR.id = ST.product_id`);
    return productsResponse.rows;
  } catch (err) {
    console.error(err);
    throw err;
  } finally {
    client.end();
  }
}

export function validateProduct(product: any): product is ProductEntity {
  return (typeof product.title === 'string' && product.title.length > 0) &&
    (!product.description || typeof product.description === 'string');
}

export async function findById(productId: string): Promise<ProductEntity> {
  const client = new Client(dbOptions);
  await client.connect();
  try {
    const queryResponse = await client.query(`
      select
        PR.*,
        coalesce(ST.count, 0) as count
      from
        public.products PR
      left join public.stocks ST on
        PR.id = ST.product_id
      where PR.id = $1`, [productId]);
    const [product] = queryResponse.rows;
    return product;
  } catch (err) {
    console.error(err);
    throw err;
  } finally {
    client.end();
  }
}

export async function createProduct(product: Partial<ProductEntity>): Promise<{
  id: string,
}> {
  const client = new Client(dbOptions);
  await client.connect();
  try {
    await client.query('BEGIN');
    const {
      title,
      description,
      price,
      count = 0,
    } = product || {};
    const createProductResponse = await client.query<ProductEntity>(`
      insert into public.products (title, description, price)
      values($1, $2, $3) returning *`, [title, description, Number(price)]);
    const [createdProduct] = createProductResponse.rows;

    const createStockResponse = await client.query<StockEntity>(`
      insert into public.stocks (product_id, count)
      values($1, $2) returning count`, [createdProduct.id, Number(count)]);
    const [createdStock] = createStockResponse.rows;
    createdProduct.count = createdStock.count;
    await client.query('COMMIT');
    return createdProduct;
  } catch (err) {
    await client.query('ROLLBACK');
    console.error(err);
    throw err;
  } finally {
    client.end();
  }
}
