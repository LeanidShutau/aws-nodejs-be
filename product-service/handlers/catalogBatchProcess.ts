import { SQSHandler } from 'aws-lambda';
import 'source-map-support/register';

import { createProduct, validateProduct } from './products-db';

export const catalogBatchProcess: SQSHandler = async (event, _context) => {
  for (const { body } of event.Records) {
    try {
      const parsedBody = JSON.parse(body);
      if (validateProduct(parsedBody)) {
        await createProduct(parsedBody);
      } else {
        console.error(`Invalid product:`, body);
      }
    } catch (err) {
      console.error(err);
    } 
  }
}
