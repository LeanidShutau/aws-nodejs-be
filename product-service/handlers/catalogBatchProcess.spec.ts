import { SQSEvent } from 'aws-lambda';
import * as AWSMock from 'aws-sdk-mock';

import { catalogBatchProcess } from './catalogBatchProcess';
import * as productsDb from './products-db';

jest.mock('./products-db');
AWSMock.mock('SNS', 'publish', 'test-message');

it('should create products', async () => {
  const products = [
    {
      title: 'Test 1',
      description: 'Description',
      count: 2,
      price: 20,
    },
    {
      title: 'Test 2',
      description: 'Description',
      count: 3,
      price: 10,
    },
  ];
  const event = ({
    Records: products.map((product) => ({
      body: JSON.stringify(product),
    })),
  } as unknown) as SQSEvent;

  (productsDb.validateProduct as any).mockResolvedValue(true);
  const createSpy = jest.spyOn(productsDb, 'createProduct'); 
  await catalogBatchProcess(event, null, null);
  expect(createSpy.mock.calls).toMatchSnapshot();
});
