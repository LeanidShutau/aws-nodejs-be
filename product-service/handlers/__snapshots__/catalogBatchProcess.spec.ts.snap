// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create products 1`] = `
Array [
  Array [
    Object {
      "count": 2,
      "description": "Description",
      "price": 20,
      "title": "Test 1",
    },
  ],
  Array [
    Object {
      "count": 3,
      "description": "Description",
      "price": 10,
      "title": "Test 2",
    },
  ],
]
`;
